File: arduinobot_ws/src/arduinobot_firmware/src/simple_serial_receiver.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <libserial/SerialPort.h>

#include <chrono>

using namespace std::chrono_literals;

//inhereting the Node class
class SimpleSerialReceiver : public rclcpp::Node{

public:
    SimpleSerialReceiver() : Node("simple_serial_receiver")
    {
        //Create the publisher with the name and number of message queue
        pub_ = create_publisher<std_msgs::msg::String>("serial_receiver",10);

        //binding the call back with the class instance,
        // and showing that the callback function is from the class
        timer_ = create_wall_timer(0.01s,std::bind(&SimpleSerialReceiver::timerCallBack,this));

        //we are declaring parameters so that we can get the port and baud rate for the arduino without hardcoding it 
        declare_parameter<std::string>("port", "/dev/ttyUSB0");
        port_ = get_parameter("port").as_string();
        arduino_.Open(port_);
        arduino_.SetBaudRate(LibSerial::BaudRate::BAUD_115200);
    }

    ~SimpleSerialReceiver()
    {
        arduino_.Close();
    }

    void timerCallBack()
    {
        //creating a message variable to send
        auto message = std_msgs::msg::String();
        if(rclcpp::ok() && arduino_.IsDataAvailable())
        {
            arduino_.ReadLine(message.data);
        }
        pub_->publish(message);
    }

private:
    //sharedptr is the pointer used to refer to the publisher and timer
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr timer_;
    LibSerial::SerialPort arduino_;
    std::string port_;
};

int main(int args, char* argv[])
{
    rclcpp::init(args,argv);
    auto node = std::make_shared<SimpleSerialReceiver>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0; 
}

File: arduinobot_ws/install/arduinobot_firmware/local/lib/python3.10/dist-packages/arduinobot_firmware/simple_serial_receiver.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import serial

class SimpleSerialReceiver(Node):
    def __init__(self):
        #initalize the constructor of the Node with the name
        super().__init__("simple_serial_receiver")

        self.declare_parameter("port", "/dev/ttyUSB0")
        self.declare_parameter("baudrate", 115200)

        self.port_ = self.get_parameter("port").value
        self.baudrate_ = self.get_parameter("baudrate").value

        #Create the publisher
        #We need to put the proper message type/Name of topic/buffer size in messages
        self.pub = self.create_publisher(String,"serial_receiver",10)
        #Frequency of messages ex: 1 message/second
        self.frequency = 0.01

        self.arduino_ = serial.Serial(port=self.port_, baudrate=self.baudrate_, timeout=0.1 )
        #it's a timer with the frequuency of 1hz where it calls the callback that we will define
        self.timer = self.create_timer(self.frequency,self.timerCallback)

    def timerCallback(self):
        #create a new message to be published
        if rclpy.ok() and self.arduino_.is_open:
            data = self.arduino_.readline()

            try:
                data.decode("utf-8")
            except:
                return
            msg = String()
            msg.data = str(data)
            self.pub.publish(msg)

def main():
    rclpy.init()
    simple_serial_receiver = SimpleSerialReceiver()
    rclpy.spin(simple_serial_receiver)
    simple_serial_receiver.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()